src/Spotflow.InMemory.Azure.ServiceBus/InMemoryServiceBusSessionProcessor.cs

```diff
@@ -0,0 +1,444 @@
+                    if (activeSessions.Count < MaxConcurrentSessions)
+                    {
+                        await _sessionConcurrencySemaphore.WaitAsync(cancellationToken);
+                        try
+                        {
+                            var sessionReceiver = await TryAcceptNextSessionAsync(cancellationToken);
+                            if (sessionReceiver != null)
+                            {
+                                var sessionTask = Task.Run(async () =>
+                                {
+                                    try
+                                    {
+                                        await ProcessSingleSessionAsync(sessionReceiver, cancellationToken);
+                                    }
+                                    finally
+                                    {
+                                        _sessionConcurrencySemaphore.Release();
+                                    }
+                                }, cancellationToken);
+                                activeSessions.TryAdd(sessionReceiver.SessionId, sessionTask);
+                            }
+                            else
+                            {
+                                _sessionConcurrencySemaphore.Release();
+                            }
+                        }
+                        catch (Exception)
+                        {
+                            _sessionConcurrencySemaphore.Release();
+                            throw;
+                        }
+                    }
+                    else
+                    {
+                        // give slack when at max capacity
+                        await Task.Delay(100, cancellationToken);
+                    }
+                }
+                catch (OperationCanceledException) when (cancellationToken.IsCancellationRequested)
+                {
+                    break;
+                }
+                catch (Exception e)
+                {
+                    await HandleErrorAsync(e, cancellationToken);
+                }
+            }
+        }
+        finally
+        {
+            if (!activeSessions.IsEmpty)
+            {
+                await Task.WhenAll(activeSessions.Values);
+            }
+        }
+    }
+    private async Task ProcessSingleSessionAsync(InMemoryServiceBusSessionReceiver sessionReceiver, CancellationToken cancellationToken)
+    {
+        var messageCallbacksSemaphore = new SemaphoreSlim(MaxConcurrentCallsPerSession, MaxConcurrentCallsPerSession);
+        var messageProcessingTasks = new List<Task>();
+
+        try
+        {
+            var initArgs = new ProcessSessionEventArgs(sessionReceiver, Identifier, cancellationToken);
+
+            await OnSessionInitializingAsync(initArgs);
+
+            var sessionIdleTimeout = SessionIdleTimeout ?? _defaultMaxWaitTime;
+            var lastActivity = DateTimeOffset.UtcNow;
```

Comment:
Please change all time-keeping code to use `TimeProvider` from `InMemoryServiceBusProvider` instead of `DateTimeOffset.UtcNow` to respect the user-provided `TimeProvider`.

Also, make sure to use monotonic clock instead of regular wall-time for this purpose: 

```
var lastActivity = timeProvider.GetTimestamp();
...
var elapsed = timeProvider.GetElapsedTime(lastActivity);
```